File: app/src/main/java/guepardoapps/mycoins/controller/NavigationController.java
package guepardoapps.mycoins.controller;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;

public class NavigationController {
    private static final String TAG = NavigationController.class.getSimpleName();
    private Logger _logger;

    private Context _context;

    public NavigationController(@NonNull Context context) {
        _logger = new Logger(TAG);
        _logger.Debug("Created new " + TAG + "...");
        _context = context;
    }

    public void NavigateTo(
            @NonNull Class<?> activity,
            boolean finish) {
        _context.startActivity(new Intent(_context, activity));
        if (finish) {
            ((Activity) _context).finish();
        }
    }

    public void NavigateWithData(
            @NonNull Class<?> activity,
            @NonNull Bundle data,
            boolean finish) {
        Intent navigateIntent = new Intent(_context, activity);
        navigateIntent.putExtras(data);
        _context.startActivity(navigateIntent);
        if (finish) {
            ((Activity) _context).finish();
        }
    }

    public boolean NavigateToOtherApplication(
            @NonNull String packageName,
            boolean finish) {
        PackageManager packageManager = _context.getPackageManager();
        try {
            Intent startAppIntent = packageManager.getLaunchIntentForPackage(packageName);

            if (startAppIntent == null) {
                return false;
            }

            startAppIntent.addCategory(Intent.CATEGORY_LAUNCHER);
            _context.startActivity(startAppIntent);

            if (finish) {
                ((Activity) _context).finish();
            }

            return true;
        } catch (Exception e) {
            _logger.Error(e.toString());
            return false;
        }
    }
}

File: app/src/main/java/guepardoapps/mycoins/controller/NavigationController.kt
package guepardoapps.mycoins.controller

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle

internal class NavigationController(val context: Context) : INavigationController {
    override fun navigate(activity: Class<*>, finish: Boolean) {
        context.startActivity(Intent(context, activity))
        if (finish) {
            (context as Activity).finish()
        }
    }

    override fun navigateWithData(activity: Class<*>, data: Bundle, finish: Boolean) {
        val navigateIntent = Intent(context, activity)
        navigateIntent.putExtras(data)
        context.startActivity(navigateIntent)
        if (finish) {
            (context as Activity).finish()
        }
    }

    override fun navigateToOtherApp(packageName: String, finish: Boolean): Boolean {
        val packageManager = context.packageManager
        try {
            val startAppIntent = packageManager.getLaunchIntentForPackage(packageName)
                    ?: return false

            startAppIntent.addCategory(Intent.CATEGORY_LAUNCHER)
            context.startActivity(startAppIntent)

            if (finish) {
                (context as Activity).finish()
            }

            return true
        } catch (e: Exception) {
            return false
        }
    }
}